
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../../Code_Snippets_Functions/">
      
      
        <link rel="next" href="../../some_day_list/The_Some_Day_List/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.12">
    
    
      
        <title>Checking Mutation Status in scRNA-seq Data - Obscure Bioinformatics Knowledge Dump</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.2afb09e1.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="pink" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#checking-mutation-status-in-scrna-seq-data" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Obscure Bioinformatics Knowledge Dump" class="md-header__button md-logo" aria-label="Obscure Bioinformatics Knowledge Dump" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Obscure Bioinformatics Knowledge Dump
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Checking Mutation Status in scRNA-seq Data
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="pink" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="pink" data-md-color-accent="blue"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Obscure Bioinformatics Knowledge Dump" class="md-nav__button md-logo" aria-label="Obscure Bioinformatics Knowledge Dump" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Obscure Bioinformatics Knowledge Dump
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Thoughts, Rambling, and Code
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../Code_Snippets_Functions/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Useful Code Snippets and Functions
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Tutorials
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Checking Mutation Status in scRNA-seq Data
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Checking Mutation Status in scRNA-seq Data
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#the-mutations" class="md-nav__link">
    <span class="md-ellipsis">
      The Mutation(s)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#the-single-cell-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      The Single Cell Dataset
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#using-vartrix" class="md-nav__link">
    <span class="md-ellipsis">
      Using Vartrix
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Using Vartrix">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#install" class="md-nav__link">
    <span class="md-ellipsis">
      Install
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running" class="md-nav__link">
    <span class="md-ellipsis">
      Running
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#using-the-output" class="md-nav__link">
    <span class="md-ellipsis">
      Using the Output
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Some day list
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Some day list
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../some_day_list/The_Some_Day_List/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    The "some day" List
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#the-mutations" class="md-nav__link">
    <span class="md-ellipsis">
      The Mutation(s)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#the-single-cell-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      The Single Cell Dataset
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#using-vartrix" class="md-nav__link">
    <span class="md-ellipsis">
      Using Vartrix
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Using Vartrix">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#install" class="md-nav__link">
    <span class="md-ellipsis">
      Install
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running" class="md-nav__link">
    <span class="md-ellipsis">
      Running
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#using-the-output" class="md-nav__link">
    <span class="md-ellipsis">
      Using the Output
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="checking-mutation-status-in-scrna-seq-data">Checking Mutation Status in scRNA-seq Data<a class="headerlink" href="#checking-mutation-status-in-scrna-seq-data" title="Permanent link">&para;</a></h1>
<p>Sometimes, it's useful to look for known mutations in your scRNA-seq reads, especially if you expect a mutation to be present in certain populations and not others, e.g. an inducible model system or a recurrent tumor versus a diagnostic one.</p>
<p>This process is relatively simple, though somewhat annoying. Let us delve into that annoyance now.</p>
<div class="admonition caution">
<p class="admonition-title">A Note of Warning</p>
<p><strong>Note that this process is subject to high amounts of ambiguity.</strong> Given the sparsity of most scRNA-seq datasets, there may not be any reads for a given variant, especially if the gene is not highly expressed. Even if there is coverage, a WT call does not mean that the record is actually WT - it may only have a single read, so the other allele could well be mutant. Point being, you can really only tell if there is any support for the alternate allele in a given cell, not that the cell is WT.</p>
</div>
<h2 id="the-mutations">The Mutation(s)<a class="headerlink" href="#the-mutations" title="Permanent link">&para;</a></h2>
<p>In this case, I only care about one mutation, so I handmade my VCF for the H3.3K27M mutation for mouse (mm10):</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>##fileformat=VCFv4.2
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>#CHROM  POS ID  REF ALT QUAL    FILTER  INFO
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>chr1    180811844   .   T   A   .   PASS    .
</code></pre></div>
<p>Ultimately, you just need a VCF of mutations derived from the same reference as your single cell data. If you have matched bulk WGS or WES data, variants called from that can be used. You <em>can</em> call variants from the scRNA-seq data itself, but it isn't recommended and won't be covered here. Use the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531192-RNAseq-short-variant-discovery-SNPs-Indels-">GATK RNA-seq variant discovery best practices</a> for that.</p>
<p>This file needs to be bgzipped and indexed via <code>tabix</code> as well.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>bgzip<span class="w"> </span>k27m.vcf
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>tabix<span class="w"> </span>k27m.vcf.gz
</code></pre></div>
<h2 id="the-single-cell-dataset">The Single Cell Dataset<a class="headerlink" href="#the-single-cell-dataset" title="Permanent link">&para;</a></h2>
<p>For your scRNA-seq dataset, you need aligned reads in BAM format, position sorted. Acquire these however you want, STARsolo, CellRanger, whatever. CellRanger BAMs will work by default.</p>
<p>You also need a list of cell barcodes in the dataset - the <code>barcodes.tsv.gz</code> file output by CellRanger works.</p>
<h2 id="using-vartrix">Using Vartrix<a class="headerlink" href="#using-vartrix" title="Permanent link">&para;</a></h2>
<p><a href="https://github.com/10XGenomics/vartrix">vartrix</a> is a tool built by 10X to genotype specific sites in scRNA-seq data for known mutations. </p>
<h3 id="install">Install<a class="headerlink" href="#install" title="Permanent link">&para;</a></h3>
<p>Via conda:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>conda<span class="w"> </span>install<span class="w"> </span>-c<span class="w"> </span>bioconda<span class="w"> </span>vartrix
</code></pre></div>
<h3 id="running">Running<a class="headerlink" href="#running" title="Permanent link">&para;</a></h3>
<p>Pretty simple to run. Ya just feed in your VCF, the bam, the genome, the barcodes, and an output name. Below is a script to submit LSF jobs for a directory containing CellRanger runs. I use the raw feature barcodes, as I am doing my own cell calling with alevin-fry counts for my downstream analyses, so I don't want any of my potential cells filtered out.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="ch">#!/bin/bash</span>
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>module<span class="w"> </span>load<span class="w"> </span>conda3/202105
<a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a>conda<span class="w"> </span>activate<span class="w"> </span>vartrix
<a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a>
<a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a><span class="k">for</span><span class="w"> </span>f<span class="w"> </span><span class="k">in</span><span class="w"> </span>./cellranger_v6/*<span class="p">;</span><span class="w"> </span><span class="k">do</span>
<a id="__codelineno-3-6" name="__codelineno-3-6" href="#__codelineno-3-6"></a><span class="w">    </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;</span><span class="nv">$f</span><span class="s2">&quot;</span><span class="p">;</span>
<a id="__codelineno-3-7" name="__codelineno-3-7" href="#__codelineno-3-7"></a><span class="w">    </span><span class="nv">samp</span><span class="o">=</span><span class="k">$(</span>basename<span class="w"> </span><span class="s2">&quot;</span><span class="nv">$f</span><span class="s2">&quot;</span><span class="k">)</span>
<a id="__codelineno-3-8" name="__codelineno-3-8" href="#__codelineno-3-8"></a>
<a id="__codelineno-3-9" name="__codelineno-3-9" href="#__codelineno-3-9"></a><span class="w">    </span>bsub<span class="w"> </span>-P<span class="w"> </span>dnb<span class="w"> </span>-J<span class="w"> </span>vartrix_<span class="nv">$samp</span><span class="w"> </span>-q<span class="w"> </span>standard<span class="w"> </span>-n<span class="w"> </span><span class="m">4</span><span class="w"> </span>-R<span class="w"> </span><span class="s2">&quot;rusage[mem=4000] span[hosts=1]&quot;</span><span class="w"> </span>vartrix<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-3-10" name="__codelineno-3-10" href="#__codelineno-3-10"></a><span class="w">                     </span>-v<span class="w"> </span>k27m.vcf.gz<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-3-11" name="__codelineno-3-11" href="#__codelineno-3-11"></a><span class="w">                     </span>-b<span class="w"> </span><span class="s2">&quot;</span><span class="nv">$f</span><span class="s2">&quot;</span>/outs/possorted_genome_bam.bam<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-3-12" name="__codelineno-3-12" href="#__codelineno-3-12"></a><span class="w">                     </span>-f<span class="w"> </span>./refdata-gex-mm10-2020-A/fasta/genome.fa<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-3-13" name="__codelineno-3-13" href="#__codelineno-3-13"></a><span class="w">                     </span>-c<span class="w"> </span><span class="s2">&quot;</span><span class="nv">$f</span><span class="s2">&quot;</span>/outs/raw_feature_bc_matrix/barcodes.tsv.gz<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-3-14" name="__codelineno-3-14" href="#__codelineno-3-14"></a><span class="w">                     </span>-o<span class="w"> </span>./vartrix/<span class="s2">&quot;</span><span class="nv">$samp</span><span class="s2">&quot;</span>.K27M_vartrix.mtx
<a id="__codelineno-3-15" name="__codelineno-3-15" href="#__codelineno-3-15"></a>
<a id="__codelineno-3-16" name="__codelineno-3-16" href="#__codelineno-3-16"></a><span class="k">done</span>
</code></pre></div>
<p>Edit as needed, change to full paths, whatever. Run the above script via <code>bash vartrix.sh</code> assuming you saved it as such. It should be run from the directory above the one containing all your CellRanger runs.</p>
<h2 id="using-the-output">Using the Output<a class="headerlink" href="#using-the-output" title="Permanent link">&para;</a></h2>
<p>This will spit out <code>.mtx</code> files for each sample, which we likely want to get into R and slapped onto our single cell object. I use Bioconductor packages, so my object is a <code>SingleCellExperiment</code>. Seurat will be similar, and the <a href="https://github.com/10XGenomics/vartrix">vartrix</a> Github has an example of how to pull stuff out. I know I only have one variant, so I just want to tack it onto my SCE's <code>colData</code>.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="c1"># Get Sample files.</span>
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a><span class="n">samp.files</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list.files</span><span class="p">(</span><span class="s">&quot;./10X_scRNA_v3/processed/vartrix&quot;</span><span class="p">,</span><span class="w"> </span>
<a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a><span class="n">recursive</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span>
<a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a>
<a id="__codelineno-4-5" name="__codelineno-4-5" href="#__codelineno-4-5"></a><span class="n">samp.ids</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;P14K-1&quot;</span><span class="p">,</span><span class="s">&quot;P14K-2&quot;</span><span class="p">,</span><span class="s">&quot;P14K-3&quot;</span><span class="p">,</span><span class="s">&quot;P14K-4&quot;</span><span class="p">,</span>
<a id="__codelineno-4-6" name="__codelineno-4-6" href="#__codelineno-4-6"></a><span class="w">              </span><span class="s">&quot;P14-1&quot;</span><span class="p">,</span><span class="s">&quot;P14-2&quot;</span><span class="p">,</span><span class="s">&quot;P14-3&quot;</span><span class="p">,</span><span class="s">&quot;P14-4&quot;</span><span class="p">,</span>
<a id="__codelineno-4-7" name="__codelineno-4-7" href="#__codelineno-4-7"></a><span class="w">              </span><span class="s">&quot;P21K-1&quot;</span><span class="p">,</span><span class="s">&quot;P21K-2&quot;</span><span class="p">,</span><span class="s">&quot;P21K-3&quot;</span><span class="p">,</span><span class="s">&quot;P21K-4&quot;</span><span class="p">,</span>
<a id="__codelineno-4-8" name="__codelineno-4-8" href="#__codelineno-4-8"></a><span class="w">              </span><span class="s">&quot;P21-1&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;P21-2&quot;</span><span class="p">,</span><span class="s">&quot;P21-3&quot;</span><span class="p">,</span><span class="s">&quot;P21-4&quot;</span><span class="p">,</span>
<a id="__codelineno-4-9" name="__codelineno-4-9" href="#__codelineno-4-9"></a><span class="w">              </span><span class="s">&quot;P4K-4&quot;</span><span class="p">,</span><span class="s">&quot;P4K-2&quot;</span><span class="p">,</span><span class="s">&quot;P4K-3&quot;</span><span class="p">,</span><span class="s">&quot;P4K-1&quot;</span><span class="p">,</span>
<a id="__codelineno-4-10" name="__codelineno-4-10" href="#__codelineno-4-10"></a><span class="w">              </span><span class="s">&quot;P4-2&quot;</span><span class="p">,</span><span class="s">&quot;P4-3&quot;</span><span class="p">,</span><span class="s">&quot;P4-1&quot;</span><span class="p">)</span>
<a id="__codelineno-4-11" name="__codelineno-4-11" href="#__codelineno-4-11"></a>
<a id="__codelineno-4-12" name="__codelineno-4-12" href="#__codelineno-4-12"></a><span class="n">first</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">TRUE</span>
<a id="__codelineno-4-13" name="__codelineno-4-13" href="#__codelineno-4-13"></a>
<a id="__codelineno-4-14" name="__codelineno-4-14" href="#__codelineno-4-14"></a><span class="n">all.k27m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">()</span>
<a id="__codelineno-4-15" name="__codelineno-4-15" href="#__codelineno-4-15"></a>
<a id="__codelineno-4-16" name="__codelineno-4-16" href="#__codelineno-4-16"></a><span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">samp</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="nf">seq_along</span><span class="p">(</span><span class="n">samp.files</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<a id="__codelineno-4-17" name="__codelineno-4-17" href="#__codelineno-4-17"></a><span class="w">  </span><span class="n">mm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">readMM</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;./10X_scRNA_v3/processed/vartrix/&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">samp.files</span><span class="p">[</span><span class="n">samp</span><span class="p">]))</span>
<a id="__codelineno-4-18" name="__codelineno-4-18" href="#__codelineno-4-18"></a><span class="w">  </span><span class="n">sbase</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">unlist</span><span class="p">(</span><span class="nf">strsplit</span><span class="p">(</span><span class="n">samp.files</span><span class="p">[</span><span class="n">samp</span><span class="p">],</span><span class="w"> </span><span class="s">&quot;.&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">fixed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))[</span><span class="m">1</span><span class="p">]</span>
<a id="__codelineno-4-19" name="__codelineno-4-19" href="#__codelineno-4-19"></a><span class="w">  </span><span class="n">bars</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.table</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;./10X_scRNA_v3/processed/cellranger_v6/&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sbase</span><span class="p">,</span><span class="w"> </span>
<a id="__codelineno-4-20" name="__codelineno-4-20" href="#__codelineno-4-20"></a><span class="w">                            </span><span class="s">&quot;/outs/raw_feature_bc_matrix/barcodes.tsv.gz&quot;</span><span class="p">),</span><span class="w"> </span>
<a id="__codelineno-4-21" name="__codelineno-4-21" href="#__codelineno-4-21"></a><span class="w">                     </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span>
<a id="__codelineno-4-22" name="__codelineno-4-22" href="#__codelineno-4-22"></a><span class="w">  </span><span class="n">mm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.data.frame</span><span class="p">(</span><span class="n">mm</span><span class="p">)</span>
<a id="__codelineno-4-23" name="__codelineno-4-23" href="#__codelineno-4-23"></a><span class="w">  </span><span class="nf">colnames</span><span class="p">(</span><span class="n">mm</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bars</span><span class="o">$</span><span class="n">V1</span>
<a id="__codelineno-4-24" name="__codelineno-4-24" href="#__codelineno-4-24"></a><span class="w">  </span><span class="nf">colnames</span><span class="p">(</span><span class="n">mm</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="n">samp.ids</span><span class="p">[</span><span class="n">samp</span><span class="p">],</span><span class="w"> </span><span class="s">&quot;_&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">colnames</span><span class="p">(</span><span class="n">mm</span><span class="p">))</span>
<a id="__codelineno-4-25" name="__codelineno-4-25" href="#__codelineno-4-25"></a>
<a id="__codelineno-4-26" name="__codelineno-4-26" href="#__codelineno-4-26"></a><span class="w">  </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="n">first</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<a id="__codelineno-4-27" name="__codelineno-4-27" href="#__codelineno-4-27"></a><span class="w">    </span><span class="n">all.k27m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mm</span>
<a id="__codelineno-4-28" name="__codelineno-4-28" href="#__codelineno-4-28"></a><span class="w">    </span><span class="n">first</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">FALSE</span>
<a id="__codelineno-4-29" name="__codelineno-4-29" href="#__codelineno-4-29"></a><span class="w">  </span><span class="p">}</span><span class="w"> </span><span class="kr">else</span><span class="w"> </span><span class="p">{</span>
<a id="__codelineno-4-30" name="__codelineno-4-30" href="#__codelineno-4-30"></a><span class="w">    </span><span class="n">all.k27m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">cbind</span><span class="p">(</span><span class="n">all.k27m</span><span class="p">,</span><span class="w"> </span><span class="n">mm</span><span class="p">)</span>
<a id="__codelineno-4-31" name="__codelineno-4-31" href="#__codelineno-4-31"></a><span class="w">  </span><span class="p">}</span>
<a id="__codelineno-4-32" name="__codelineno-4-32" href="#__codelineno-4-32"></a><span class="p">}</span>
</code></pre></div>
<p>Check that all our cells in our <code>SCE</code> object named <code>sce</code> are found.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>table(colnames(sce) %in% colnames(all.k27m))
</code></pre></div>
<p>Which they are. So now add the calls to the metadata.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>rownames(all.k27m) &lt;- &quot;chr1_180811844&quot;
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a>all.k27m &lt;- all.k27m[,colnames(sce)]
<a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a>all.k27m &lt;- as.data.frame(t(all.k27m))
<a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a>
<a id="__codelineno-6-5" name="__codelineno-6-5" href="#__codelineno-6-5"></a># No reads detected
<a id="__codelineno-6-6" name="__codelineno-6-6" href="#__codelineno-6-6"></a>all.k27m$V1 &lt;- stringr::str_replace(as.character(all.k27m$V1), &quot;0&quot;, &quot;Unknown&quot;)
<a id="__codelineno-6-7" name="__codelineno-6-7" href="#__codelineno-6-7"></a># Only ref detected
<a id="__codelineno-6-8" name="__codelineno-6-8" href="#__codelineno-6-8"></a>all.k27m$V1 &lt;- stringr::str_replace(as.character(all.k27m$V1), &quot;1&quot;, &quot;Ref_Only&quot;)
<a id="__codelineno-6-9" name="__codelineno-6-9" href="#__codelineno-6-9"></a># Only alt detected
<a id="__codelineno-6-10" name="__codelineno-6-10" href="#__codelineno-6-10"></a>all.k27m$V1 &lt;- stringr::str_replace(as.character(all.k27m$V1), &quot;2&quot;, &quot;Mut_Only&quot;)
<a id="__codelineno-6-11" name="__codelineno-6-11" href="#__codelineno-6-11"></a># Both alleles detected
<a id="__codelineno-6-12" name="__codelineno-6-12" href="#__codelineno-6-12"></a>all.k27m$V1 &lt;- stringr::str_replace(as.character(all.k27m$V1), &quot;3&quot;, &quot;Mut_WT&quot;)
<a id="__codelineno-6-13" name="__codelineno-6-13" href="#__codelineno-6-13"></a>
<a id="__codelineno-6-14" name="__codelineno-6-14" href="#__codelineno-6-14"></a>sce$H3K27M_gt &lt;- all.k27m$V1
</code></pre></div>
<p>Great, now my <code>SCE</code> contains a column containing the genotype info, so we can plot our calls. </p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a><span class="nf">dittoDimPlot</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;H3K27M_gt&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">reduction.use</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;UMAP_m.dist0.2_n.neigh10&quot;</span><span class="p">),</span><span class="w"> </span>
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a><span class="w">             </span><span class="n">do.raster</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">color.panel</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;red&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;orange&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;black&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;grey80&quot;</span><span class="p">),</span><span class="w"> </span>
<a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a><span class="w">             </span><span class="n">order</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;decreasing&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">theme</span><span class="p">(</span><span class="n">aspect.ratio</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span>
</code></pre></div>
<p><img alt="Pasted image 20220509130339.png" src="../../attachments/Pasted%20image%2020220509130339.png" /></p>
<p>Looks pretty decent. See a few pops that didn't get much induction, and some that did, which was to be expected.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2022 Jared Andrews
    </div>
  
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": [], "search": "../../assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.c8b220af.min.js"></script>
      
    
  </body>
</html>